### PR TITLE
Check beatmap cover images on qualification

### PR DESCRIPTION
**Pull request by nekodex**
...and queue a regeneration if any are missing

### CODE REVIEW

File: app/Models/Beatmapset.php [hunk 1]
```diff
@@ -830,6 +831,32 @@ public function regenerateCovers(array $sizesToRegenerate = null)
...
+                curl_exec($ch);
+                $responseCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+                curl_close($ch);
+
+                if ($responseCode !== 200) {
+                    $allGood = false;
+                    break;
+                }
+            }
+        } catch (\Exception $e) {
```
Comment (nanaya) [inline app/Models/Beatmapset.php]: what is this catching?

File: app/Models/Beatmapset.php [hunk 2]
```diff
@@ -830,6 +831,32 @@ public function regenerateCovers(array $sizesToRegenerate = null)
         $this->update(['cover_updated_at' => $this->freshTimestamp()]);
     }
 
+    public function allCoverImagesPresent()
+    {
+        $allGood = true;
+        try {
+            foreach ($this->allCoverURLs() as $size => $url) {
+                $ch = curl_init($url);
+                curl_setopt($ch, CURLOPT_HEADER, true);
```
Comment (nanaya) [inline app/Models/Beatmapset.php]: there's this `curl_setopt_array`.

File: app/Jobs/CheckBeatmapsetCovers.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * Execute the job.
+     *
+     * @return void
+     */
+    public function handle()
+    {
+        // trigger a cover regeneration if any cover images are missing
+        if (!$this->beatmapset->allCoverImagesPresent()) {
+            $job = (new RegenerateBeatmapsetCover($this->beatmapset))->onQueue('beatmap_high');
```
Comment (nanaya) [inline app/Jobs/CheckBeatmapsetCovers.php]: Apparently there's `Dispatchable` trait which may simplify things.

File: app/Models/Beatmapset.php [hunk 4]
```diff
@@ -830,6 +831,32 @@ public function regenerateCovers(array $sizesToRegenerate = null)
         $this->update(['cover_updated_at' => $this->freshTimestamp()]);
     }
 
+    public function allCoverImagesPresent()
+    {
+        $allGood = true;
+        try {
+            foreach ($this->allCoverURLs() as $size => $url) {
+                $ch = curl_init($url);
```
Comment (nanaya) [inline app/Models/Beatmapset.php]: better off as helper function?

File: app/Models/Beatmapset.php [hunk 5]
```diff
@@ -830,6 +831,32 @@ public function regenerateCovers(array $sizesToRegenerate = null)
         $this->update(['cover_updated_at' => $this->freshTimestamp()]);
     }
 
+    public function allCoverImagesPresent()
+    {
+        $allGood = true;
```
Comment (nanaya) [inline app/Models/Beatmapset.php]: a bit shorter...

```php
... function ...()
{
    foreach (... as ...) {
        if (!url_check(...)) {
            return false;
        }
    }

    return true;
}
```

?

File: app/Jobs/CheckBeatmapsetCovers.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+/**
+ *    Copyright 2015-2017 ppy Pty. Ltd.
```
Comment (nanaya) [inline app/Jobs/CheckBeatmapsetCovers.php]: 2018 =D

File: app/Jobs/CheckBeatmapsetCovers.php (hunk 3 above)
Comment (nekodex) [reply app/Jobs/CheckBeatmapsetCovers.php]: Yeah I don't think it really helps... if anything, I'd argue it makes it less readable not more readable?

```php
$job = (new RegenerateBeatmapsetCover($this->beatmapset))->onQueue('beatmap_high');
dispatch($job);
```
vs
```php
(new RegenerateBeatmapsetCover($this->beatmapset))->dispatch()->onQueue('beatmap_high');
```

Comment (notbakaneko) [reply app/Jobs/CheckBeatmapsetCovers.php]: I don't think `Dispatchable` helps when the job is a different class.

Comment (nanaya) [reply app/Jobs/CheckBeatmapsetCovers.php]: more like

```
RegenerateBeatmapsetCover::dispatch($this->beatmapset))->onQueue('beatmap_high');
```

Comment (nekodex) [reply app/Jobs/CheckBeatmapsetCovers.php]: My readability complaint remains, `dispatch()->onQueue()` makes no sense to me, I expect dispatch to... dispatch, not to be chained onto to enqueue elsewhere

**Pull request merged**